File: pkg/textwrap/textwrap_test.go
// SPDX-FileCopyrightText: (c) 2026 Rafal Zajac
// SPDX-License-Identifier: MIT

package textwrap

import (
	"testing"

	"github.com/ctx42/goldkit/pkg/goldkit"
	"github.com/ctx42/testing/pkg/assert"
	"github.com/ctx42/testing/pkg/must"
)

func Test_Wrap_tabular(t *testing.T) {
	tt := []struct {
		testN string
		file  string
	}{
		{"basic reflow", "testdata/basic.yml"},
		{"hyphenated word stays intact", "testdata/hyphenated.yml"},
		{"over-long token overflows its line", "testdata/overflow.yml"},
		{"double-width runes", "testdata/double_width.yml"},
		{"whitespace collapses", "testdata/collapse.yml"},
		{"width zero means no limit", "testdata/no_limit.yml"},
		{"empty input", "testdata/empty.yml"},
		{"all whitespace input", "testdata/whitespace.yml"},
	}

	for _, tc := range tt {
		t.Run(tc.testN, func(t *testing.T) {
			// --- Given ---
			gld := goldkit.Create(t, tc.file, nil)
			input := must.Value(gld.M().MetaGetString("input"))
			width := must.Value(gld.M().MetaGetInt("width"))

			// --- When ---
			have := Wrap(input, width)

			// --- Then ---
			gld.Assert([]byte(have))
		})
	}
}

func Test_WrapTokens_tabular(t *testing.T) {
	tt := []struct {
		testN string
		words []string
		width int
		want  string
	}{
		{
			"a token keeps its inner spaces intact",
			[]string{"see", "[Asset Data](url)", "now"},
			14,
			"see\n[Asset Data](url)\nnow",
		},
		{
			"an over-long token overflows its own line",
			[]string{"a", "[very long label](url)", "b"},
			5,
			"a\n[very long label](url)\nb",
		},
		{
			"width zero means no limit",
			[]string{"a", "b c", "d"},
			0,
			"a b c d",
		},
		{
			"negative width means no limit",
			[]string{"a", "b c", "d"},
			-1,
			"a b c d",
		},
		{
			"no words yields empty",
			nil,
			10,
			"",
		},
	}

	for _, tc := range tt {
		t.Run(tc.testN, func(t *testing.T) {
			// --- When ---
			have := WrapTokens(tc.words, tc.width)

			// --- Then ---
			assert.Equal(t, tc.want, have)
		})
	}
}

File: packages/core/test/textwrap/textwrap.test.ts
// SPDX-FileCopyrightText: (c) 2026 Rafal Zajac
// SPDX-License-Identifier: MIT

// Ported 1:1 from pkg/textwrap/textwrap_test.go. The tabular `Wrap` cases read
// the 8 golden fixtures reused verbatim from the Go suite.

import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";
import { wrap, wrapTokens } from "../../src/textwrap/textwrap.ts";
import { loadGolden, metaInt, metaString } from "../support/golden.ts";

const TESTDATA = fileURLToPath(new URL("./testdata/", import.meta.url));

describe("wrap (golden fixtures)", () => {
    const cases: ReadonlyArray<readonly [string, string]> = [
        ["basic reflow", "basic.yml"],
        ["hyphenated word stays intact", "hyphenated.yml"],
        ["over-long token overflows its line", "overflow.yml"],
        ["double-width runes", "double_width.yml"],
        ["whitespace collapses", "collapse.yml"],
        ["width zero means no limit", "no_limit.yml"],
        ["empty input", "empty.yml"],
        ["all whitespace input", "whitespace.yml"],
    ];

    it.each(cases)("%s", (_name, file) => {
        const golden = loadGolden(TESTDATA + file);

        const have = wrap(
            metaString(golden, "input"),
            metaInt(golden, "width"),
        );

        expect(have).toBe(golden.body);
    });
});

describe("wrapTokens", () => {
    const cases: ReadonlyArray<{
        name: string;
        words: string[];
        width: number;
        want: string;
    }> = [
        {
            name: "a token keeps its inner spaces intact",
            words: ["see", "[Asset Data](url)", "now"],
            width: 14,
            want: "see\n[Asset Data](url)\nnow",
        },
        {
            name: "an over-long token overflows its own line",
            words: ["a", "[very long label](url)", "b"],
            width: 5,
            want: "a\n[very long label](url)\nb",
        },
        {
            name: "width zero means no limit",
            words: ["a", "b c", "d"],
            width: 0,
            want: "a b c d",
        },
        {
            name: "negative width means no limit",
            words: ["a", "b c", "d"],
            width: -1,
            want: "a b c d",
        },
        {
            name: "no words yields empty",
            words: [],
            width: 10,
            want: "",
        },
    ];

    it.each(cases)("$name", ({ words, width, want }) => {
        expect(wrapTokens(words, width)).toBe(want);
    });
});
